search for ForeignSecurityPrincipals already implemented

d34dl0ckk opened the pull request:
Hello, 

Thanks a lot for this great tool! 

While I was working on group scopes and memberships, I needed to get the list of external/forest trust members of a group in a domain and I used `ldapsearch-ad` for that. As it uses ldap3 to handle requests, I saw that it was possible to specify the port number and have the requests addressed the global catalog instead, which contains some useful information about external membership in domain local groups (very minimum actually but still useful for Recon). 

Here is the way I did to retrieve the information hoping it is useful. 

Best regards,

d34dl0ckk committed:
Update README.md

d34dl0ckk committed:
Added search for ForeignSecurityPrincipals

d34dl0ckk committed:
minor fix

d34dl0ckk committed:
Get external Security Principals members

d34dl0ckk committed:
Update USAGE.md

yaap7 commented:
Hello @d34dl0ckk ,
Thanks for implementing this great functionnality!
I saw 5 commits on 2 files (readme.md and usage.md), but I don't see any modification in the python code itself.
Do you have a working code to implement this functionnality, or do you want me to implement it?
In the later, I will need the exact ldap filter to request on the global catalog. Could you please provide it?
Thanks again for your contribution :)

d34dl0ckk commented:
> Hello @d34dl0ckk , Thanks for implementing this great functionnality! I saw 5 commits on 2 files (readme.md and usage.md), but I don't see any modification in the python code itself. Do you have a working code to implement this functionnality, or do you want me to implement it? In the later, I will need the exact ldap filter to request on the global catalog. Could you please provide it? Thanks again for your contribution :)

Hello @yaap7,
Thanks a lot for the quick reply!

As stated in the PR title, the functionality is already available in ldapsearch-ad without further implementation, using the -t search, the proper ldap filter (objectclass=foreignSecurityPrincipal) and specifying the global catalog port (3268) along with the IP address of the DC that has this role in the forest (by default the root domain has it and data is replicated to it). 
To redirect requests to the GC instead of LDAP I just specified the port as stated here on [ldap3](https://ldap3.readthedocs.io/en/latest/server.html?highlight=port#server-object).

`$ ./ldapsearch-ad.py -l <GC_IP>:3268 -d <Domain> -u <Username> -p '<Password>' -t search -s '(objectClass=foreignSecurityPrincipal)'`

Please find an example of the output in the following.

Get the SID of a foreign user added from a domain in the trusted forest to a domain in the current (trusting) forest: 

![foreignuserusingldapsearch-ad](https://user-images.githubusercontent.com/15831888/220322907-9997d9f2-4d0e-480a-be92-28e4b270b268.png)

As you can see, at the end of the output we can see the SID of the foreign user and thus use it for further Recon later to find the group itself.

This has been run on a lab with two forests having a bidirectional forest trust relationship (Please find more details about the topology on [GOADv2](https://mayfly277.github.io/posts/GOADv2/)).

Do you want me to implement this as a new type (-t) the same way member-of is done; so that it will be directly available as a functionality? Edit: currently working on it!

Thank you! 
Best regards,

d34dl0ckk committed:
Added search-foreign-security-principals type

d34dl0ckk committed:
fixed print foreign sp with attributes

d34dl0ckk commented:
Hello @yaap7,
I just implemented the new functionality, please feel free to give feedback and I'll modify :). 

Here is the output it gives.

Either we don't specify the port number (with -n, I added this option to be able to request the GC if needed as it is faster to reply). 
![search-foreign-sp-ldaps](https://user-images.githubusercontent.com/15831888/220357442-86c4017f-cafb-448c-8deb-4c303f6631f4.png)

Or, specifying the port number (-n) and retrieve information from the GC.
![search-foreign-sp-gc](https://user-images.githubusercontent.com/15831888/220357763-4e71c87e-c187-4ba2-b73c-d656a8978f53.png)

(Tests are done on the same lab environment as previous ones).

d34dl0ckk committed:
added search-foreign-security-principals to USAGE

yaap7 commented:
Thank for your so fastcode update!
I quickly reviewed your code and it seems great.
I will test it in a few days and accept your PR as soon as possible.

yaap7 commented:
I finally merged your work into the `master` branch.
Thanks again for your proposal and sorry for the late reply.
